src/main/java/org/assertj/core/internal/Objects.java

```diff
@@ -802,6 +804,37 @@ public boolean areEqualToComparingOnlyGivenFields(Object actual, Object other,
     return PropertyOrFieldSupport.EXTRACTION.getValueOf(name, actual);
   }
 
+  /**
+   * Asserts that the actual object has the same hashCode as the given object.
+   *
+   * @param info contains information about the assertion.
+   * @param actual the given object.
+   * @param other the object to check hashCode against.
+   *
+   * @throws AssertionError if the actual object is null.
+   * @throws AssertionError if the given object is null.
+   * @throws AssertionError if the actual object has not the same hashCode as the given object.
+   */
+  public <A> void assertHasSameHashCodeAs(AssertionInfo info, A actual, Object other) {
+    assertNotNull(info, actual);
+
+    if (other == null) {
+      throw failures.failure(info, otherShouldNotBeNull());
+    }
+
+    if (actual.hashCode() != other.hashCode()) {
+      throw failures.failure(info, shouldHaveSameHashCode(actual, other));
+    }
+  }
+
+  private BasicErrorMessageFactory otherShouldNotBeNull() {
```

Comment:
put that in a `org.assertj.core.error.OtherShouldNotBeNull` similar to `ShouldNotBeNull`.